File: android/app/src/main/java/com/example/android/models/Message.kt
package com.example.android.models

import android.util.Log
import com.google.gson.annotations.SerializedName
import java.text.SimpleDateFormat
import java.util.*

data class Message(
    @SerializedName("id")
    val id: Int,
    @SerializedName("content")
    var content: String,
    @SerializedName("date")
    val date: Date,
    @SerializedName("creator")
    val creator: User,
    @SerializedName("sender")
    val sender: User,
    @SerializedName("chat")
    val chat: Int,
    @SerializedName("message_type")
    var messageType: Int,
    @SerializedName("is_sender")
    val isSender: Boolean
){
    fun getInitials(): String{
        if (sender.profile == null){
            return "U${sender.id}"
        }
        var initials = ""
        if (sender.profile!!.firstName.isNotEmpty()){
            initials += sender.profile!!.firstName[0]
        }
        if (sender.profile!!.lastName.isNotEmpty()){
            initials += sender.profile!!.lastName[0]
        }
        if (initials.isEmpty()){
            return "U${sender.id}"
        }
        return initials
    }

    fun getOwnerContent(): String{
        if (sender.profile == null) return content
        return "${sender.profile!!.firstName}: $content"
    }
    fun getTime(isEdit: Boolean = false): String{
        var time = ""
        if (isEdit && messageType == 3){
            time += "edited "
        }
        return time + SimpleDateFormat("HH:mm", Locale.ROOT).format(date)
    }
    fun getForwardedBannerText(): String{
        return "Forwarded from ${creator.name()}"
    }
}

File: backend/core/serializers/message.py
from rest_framework import serializers

from core.models import Message
from user.models import User, Profile


class ProfileSerializer(serializers.ModelSerializer):
    class Meta:
        model = Profile
        fields = ['first_name', 'last_name']


class UserSerializer(serializers.ModelSerializer):
    profile = ProfileSerializer(read_only=True)

    class Meta:
        model = User
        fields = ['id', 'profile']


class MessageResponseSerializer(serializers.ModelSerializer):
    creator = UserSerializer(read_only=True)
    sender = UserSerializer(read_only=True)
    is_sender = serializers.BooleanField(read_only=True)

    class Meta:
        model = Message
        fields = ['id', 'content', 'creator', 'sender', 'date', 'chat', 'message_type', 'is_sender']


class MessageSerializer(serializers.ModelSerializer):
    class Meta:
        model = Message
        fields = '__all__'
